feat: ES Module

babblebey opened the pull request:
This PR converts the plugin to ESM only.

Related to https://github.com/semantic-release/semantic-release/issues/2133

BREAKING CHANGE: `@semantic-release/git` is now a native ES Module. It has named exports for each plugin hook (verifyConditions, prepare)

BREAKING CHANGE: the minimum required version of semantic-release to use `@semantic-release/git` is now [v20.1.0](https://github.com/semantic-release/semantic-release/releases/tag/v20.1.0)

#### Other Notable Changes

- `@semantic-release/error` `^3` → `^4` (ESM-only)
- `aggregate-error` `^3` → `^5` (ESM-only, uses native AggregateError)
- `execa` `^5` → `^9` (ESM-only, named export)
- `p-reduce` `^2` → `^3` (ESM-only)
- Removed CJS-only devDeps: `clear-module`, `file-url`, `fs-extra`, `get-stream`, `nyc`
- Added `c8` for coverage; bumped `ava`, `sinon`, `tempy`, `semantic-release`, `prettier`
- Engine: `^22.14.0 || >= 24.10.0`; peer dep: `>=24.1.0`
- Scripts: `nyc` → `c8`, added `pretest`, removed `test:ci`

### Related Issue

- https://github.com/semantic-release/semantic-release/issues/2133
- https://github.com/semantic-release/semantic-release/issues/4032
- Closes #491
- Fixes #490

babblebey committed:
refactor: Update code to use ES6 module syntax

This commit updates the code to use ES6 module syntax by replacing the `require` statements with `import` statements. This improves the code readability and maintainability.

Refactor the `index.js` file to use ES6 module syntax

Refactor the `lib/definitions/errors.js` file to use ES6 module syntax

Refactor the `lib/get-error.js` file to use ES6 module syntax

Refactor the `lib/git.js` file to use ES6 module syntax

Refactor the `lib/prepare.js` file to use ES6 module syntax

Refactor the `lib/resolve-config.js` file to use ES6 module syntax

Refactor the `lib/verify.js` file to use ES6 module syntax

babblebey committed:
refactor: replace `lodash` with `lodash-es`

babblebey committed:
refactor: modify `"package.json".homepage` import; remove `#readme` from `"package.json".homepage`;

babblebey committed:
fix: add missing `function` keyword

babblebey committed:
refactor: adapt `debug` import in `git.js` and `prepare.js`

babblebey committed:
refactor: replace `xo` with `prettier`

babblebey committed:
Merge branch 'master' into esm

babblebey committed:
chore: update dependencies and devDependencies; improve Node.js engine support

- Bump @semantic-release/error from ^3.0.0 to ^4.0.0
- Bump aggregate-error from ^3.0.0 to ^5.0.0
- Bump execa from ^5.0.0 to ^9.0.0
- Bump p-reduce from ^2.0.0 to ^3.0.0
- Update devDependencies:
  - Bump ava from 3.15.0 to 6.2.0
  - Add c8 for coverage reporting
  - Bump prettier from ^3.3.3 to 3.5.3
  - Bump semantic-release from 21.1.2 to 24.2.1
  - Bump sinon from 17.0.1 to 19.0.2
  - Bump tempy from 1.0.1 to 3.1.0
- Update Node.js engine support to ^22.14.0 || >= 24.10.0
- Add pretest script to run linting before tests
- Replace nyc with c8 for test coverage reporting
- Update peerDependencies for semantic-release to >=24.1.0

babblebey committed:
refactor: migrate to ES modules and improve code readability in git-utils.js

babblebey committed:
refactor: migrate tests to use ES module syntax and improve readability

- Changed require statements to import statements in integration, prepare, and verify tests.
- Replaced fs-extra's outputFile with native fs/promises methods for file operations.
- Updated test cases to enhance consistency in formatting and structure.
- Improved error handling in verification tests for assets and message options.

babblebey committed:
chore: update GitHub Actions workflows to use latest action versions and improve Node.js compatibility

babblebey committed:
refactor: standardize code formatting and improve readability across multiple files

babblebey committed:
chore: enhance Node.js compatibility check by cleaning npm cache before running ls-engines

babblebey committed:
chore: update dependencies and improve test scripts

- Updated semantic-release to version ^25.0.3
- Updated sinon to version ^21.0.1
- Added npm-run-all to manage parallel test execution
- Modified test script to run linting and tests in parallel
- Introduced separate test scripts for unit and integration tests

babblebey committed:
chore: update Node.js versions and improve workflow steps in test.yml

babblebey committed:
chore: add .nvmrc file to specify Node.js version

babblebey committed:
chore: update peer dependency for semantic-release to allow earlier versions

babblebey committed:
chore: exclude integration tests from AVA test files

babblebey committed:
test: update AVA configuration for integration tests and adjust timeout settings

travi commented on .github/workflows/test.yml:
```diff
@@ -13,35 +13,45 @@ jobs:
+      - run: corepack npm audit signatures
+      - run: npm test
+
+  test:
+    runs-on: ubuntu-latest
+    needs:
+      - test_dev
+      - test_matrix
+    if: ${{ !cancelled() }}
+    steps:
+      - name: All matrix versions passed
+        if: ${{ !(contains(needs.*.result, 'failure')) }}
+        run: exit 0
+      - name: Some matrix version failed
+        if: ${{ contains(needs.*.result, 'failure') }}
+        run: exit 1
```
this is a good example to follow for the changelog update

babblebey replied:
Very well,

Attn to the failing workflows, I don't seem to understand them... A little light there please 🤔

travi commented on lib/definitions/errors.js:
```diff
@@ -1,20 +1,22 @@
+import { createRequire } from "node:module";
+
+const require = createRequire(import.meta.url);
+const { homepage } = require("../../package.json");
```
lets avoid `createRequire` here like the changelog change

travi reviewed (changes requested):
i just spot the one adjustment. then i think this is good to go. thanks a ton for the investment into getting these updated!

babblebey committed:
refactor: simplify homepage retrieval in error definitions

travi commented:
> Attn to the failing workflows, I don't seem to understand them... A little light there please 🤔

it looks like execa is making a true git call, which i would be surprised if it was doing that before. i havent looked deep enough to confirm that or what might need to change. my first thought was to check if there is an execa update available that was unblocked by the esm change, but it looks like youve already updated to the latest major. maybe there was a breaking change, or even just the nature or the esm change, that changed the way that was previously stubbed to allow the call through when it wasnt doing that before

babblebey committed:
fix(test): set default git user configuration for test repositories

babblebey committed:
fix(git): set default user configuration for shallow clone and detached head

babblebey commented:
> it looks like execa is making a true git call, which i would be surprised if it was doing that before.

I don't know how the tests have been passing originally.... but I just found some error relating to the CI runner not having a global git `user.name`/`user.email` configured, and the test helpers don't mock them before repo operations.

babblebey committed:
chore: update dependencies and enhance linting scripts

- Added lockfile-lint, ls-engines, and publint for improved linting and validation.
- Updated linting scripts to include separate commands for prettier and lockfile checks.
- Configured lockfile-lint settings for package-lock.json.

travi commented on test/helpers/git-utils.js:
```diff
@@ -142,12 +192,14 @@ async function gitShallowClone(repositoryUrl, branch = 'master', depth = 1) {
+  await execa("git", ["remote", "add", "origin", repositoryUrl], { cwd });
+  await execa("git", ["fetch", repositoryUrl], { cwd });
+  await execa("git", ["checkout", head], { cwd });
+  await execa("git", ["config", "user.email", "test@test.com"], { cwd });
+  await execa("git", ["config", "user.name", "Test"], { cwd });
```
while i understand that this gets the tests working, are we certain this doesnt impact your personal config when running locally? if it is isolated from impacting that, can you help me understand how?

babblebey replied:
Yup this doesn't mess with personal git config!

These commands only set repository-local git config in the temporary test repo (via cwd),  they do not touch the global config. Since we do not pass `--global` flag to the execa commant, personal git configs are not modified.

travi commented on package.json:
```diff
@@ -16,36 +20,37 @@
   },
   "engines": {
-    "node": ">=14.17"
+    "node": "^22.14.0 || >= 24.10.0"
```
maybe one last change for this, but this could be after merging to beta. i've been finding some packages needing even higher node version minimums, so i've been using `^22.22.2 || >=24.15` lately

babblebey replied:
> so i've been using `^22.22.2 || >=24.15` lately

Very well, Kindly list'em up in an issue or discuss, so we can appropriate
